state/watcher.go

```diff
@@ -797,6 +797,106 @@ func (w *minUnitsWatcher) Changes() <-chan []string {
 	return w.out
 }
 
+// modelMachinesProfileWatcher notifies about charm upgrade changes where a
+// machine's LXD profile may need to be changed. At startup, the watcher
+// gathers current values for a machine's UpgradeLXDProfileCharmURL, no events
+// are returned. Events are generated when there are changes to a machine's
+// UpgradeLXDProfileCharmURL.
+type modelMachinesProfileWatcher struct {
+	commonWatcher
+	known map[string]string
+	out   chan []string
+}
+
+var _ Watcher = (*modelMachinesProfileWatcher)(nil)
+
+func newModelMachinesProfileWatcher(backend modelBackend) StringsWatcher {
+	w := &modelMachinesProfileWatcher{
+		commonWatcher: newCommonWatcher(backend),
+		known:         make(map[string]string),
+		out:           make(chan []string),
+	}
+	w.tomb.Go(func() error {
+		defer close(w.out)
+		return w.loop()
+	})
+	return w
+}
+
+// WatchModelMachinesCharmProfiles returns a StringsWatcher that notifies of
+// changes to the upgrade charm profile charm url for a machine.
+func (st *State) WatchModelMachinesCharmProfiles() StringsWatcher {
+	return newModelMachinesProfileWatcher(st)
+}
+
+func (w *modelMachinesProfileWatcher) initial() (set.Strings, error) {
+	machineIds := make(set.Strings)
+	var doc machineDoc
+	newMachines, closer := w.db.GetCollection(machinesC)
+	defer closer()
+
+	iter := newMachines.Find(nil).Iter()
+	for iter.Next(&doc) {
+		w.known[doc.Id] = doc.UpgradeCharmProfileCharmURL
+		machineIds.Add(doc.Id)
+	}
+	return machineIds, iter.Close()
+}
+
+func (w *modelMachinesProfileWatcher) merge(machineIds set.Strings, change watcher.Change) error {
+	machineId := w.backend.localID(change.Id.(string))
+	if change.Revno == -1 {
+		delete(w.known, machineId)
+		machineIds.Remove(machineId)
+		return nil
+	}
+	doc := machineDoc{}
+	newMachines, closer := w.db.GetCollection(machinesC)
+	defer closer()
+	if err := newMachines.FindId(change.Id).One(&doc); err != nil {
+		return err
+	}
+	profileApplication, _ := w.known[machineId]
+	w.known[machineId] = doc.UpgradeCharmProfileCharmURL
+	if doc.UpgradeCharmProfileCharmURL != profileApplication {
+		machineIds.Add(machineId)
+	}
+	return nil
+}
+
+func (w *modelMachinesProfileWatcher) loop() (err error) {
+	ch := make(chan watcher.Change)
+	w.watcher.WatchCollectionWithFilter(machinesC, ch, isLocalID(w.backend))
+	defer w.watcher.UnwatchCollection(machinesC, ch)
+	machineIds, err := w.initial()
```

Comment:
this `err` will conflict with `(err error)`. We should either rename this `err` or create a separate variable for `machineIds` and drop `:=` i.e. 

```go
var machineIds set.Strings
machineIds, err = w.initial()
...
```

Reply:
changed '(err error)' to 'error'